invertmsrc_test fails to compile due to (invalid) implicit complex -> real downconversion

```
/users/bartek/code/2019_08_27/quda_develop/tests/invertmsrc_test.cpp: In function 'int main(int, char**)':
/users/bartek/code/2019_08_27/quda_develop/tests/invertmsrc_test.cpp:553:78: error: cannot convert '__complex__ double' to 'double' in assignment
           kappa_b[xs] = 1.0/(2*(inv_param.b_5[xs]*(4.0 + inv_param.m5) + 1.0));
                                                                              ^
/users/bartek/code/2019_08_27/quda_develop/tests/invertmsrc_test.cpp:554:78: error: cannot convert '__complex__ double' to 'double' in assignment
           kappa_c[xs] = 1.0/(2*(inv_param.c_5[xs]*(4.0 + inv_param.m5) - 1.0));
                                                                              ^
/users/bartek/code/2019_08_27/quda_develop/tests/invertmsrc_test.cpp:556:181: error: cannot convert 'double*' to '__complex__ double*' for argument '4' to 'void mdw_matpc(void*, void**, void*, __complex__ double*, __complex__ double*, QudaMatPCType, int, QudaPrecision, QudaGaugeParam&, double, __complex__ double*, __complex__ double*)'
         mdw_matpc(spinorCheck[i], gauge, spinorOutMulti[i], kappa_b, kappa_c, inv_param.matpc_type, 0, inv_param.cpu_prec, gauge_param, inv_param.mass, inv_param.b_5, inv_param.c_5);
```

--- a/include/kernels/dslash_domain_wall_4d.cuh
+++ b/include/kernels/dslash_domain_wall_4d.cuh
@@ -7,26 +7,25 @@ namespace quda
 
   constexpr int size = 4096;
   static __constant__ char mobius_d[size]; // constant buffer used for Mobius coefficients for GPU kernel
-  static void *mobius_h;                   // constant buffer used for Mobius coefficients for CPU kernel
-
-  /**
-     @brief Helper function for grabbing the constant struct, whether
-     we are on the GPU or CPU.
-  */
-  template <typename real> inline __device__ __host__ complex<real> a_5(int s)
-  {
-#ifdef __CUDA_ARCH__
-    return reinterpret_cast<const complex<real> *>(mobius_d)[s];
-#else
-    return reinterpret_cast<const complex<real> *>(mobius_h)[s];
-#endif
-  }
 
   template <typename Float, int nColor, QudaReconstructType reconstruct_>
   struct DomainWall4DArg : WilsonArg<Float, nColor, reconstruct_> {
     typedef typename mapper<Float>::type real;
     int Ls;                             /** fifth dimension length */
     complex<real> a_5[QUDA_MAX_DWF_LS]; /** xpay scale factor for each 4-d subvolume */
+
+    /**
+       @brief Helper function for grabbing the constant struct, whether
+       we are on the GPU or CPU.
+    */
+    inline __device__ __host__ complex<real> a5(int s)
+    {
+#ifdef __CUDA_ARCH__
+      return reinterpret_cast<const complex<real> *>(mobius_d)[s];
+#else
+      return a_5[s];
+#endif
+    }
 
     DomainWall4DArg(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, double a, double m_5,
         const Complex *b_5, const Complex *c_5, bool xpay, const ColorSpinorField &x, int parity, bool dagger,
@@ -38,7 +37,6 @@ namespace quda
         for (int s = 0; s < Ls; s++) a_5[s] = a; // 4-d Shamir
       else
         for (int s = 0; s < Ls; s++) a_5[s] = 0.5 * a / (b_5[s] * (m_5 + 4.0) + 1.0); // 4-d Mobius
-      mobius_h = a_5;
     }
   };
 
@@ -62,10 +60,10 @@ namespace quda
     int xs = x_cb + s * arg.dc.volume_4d_cb;
     if (xpay && kernel_type == INTERIOR_KERNEL) {
       Vector x = arg.x(xs, my_spinor_parity);
-      out = x + a_5<real>(s) * out;
+      out = x + arg.a5(s) * out;
     } else if (kernel_type != INTERIOR_KERNEL && active) {
       Vector x = arg.out(xs, my_spinor_parity);
-      out = x + (xpay ? a_5<real>(s) * out : out);
+      out = x + (xpay ? arg.a5(s) * out : out);
     }
 
     if (kernel_type != EXTERIOR_KERNEL_ALL || active) arg.out(xs, my_spinor_parity) = out;

--- a/include/kernels/dslash_domain_wall_m5.cuh
+++ b/include/kernels/dslash_domain_wall_m5.cuh
@@ -16,24 +16,21 @@ namespace quda
   };
 
   constexpr int size = 4096;
-  static __constant__ char mobius_d[size]; // constant buffer used for Mobius coefficients for GPU kernel
-  static char mobius_h[size];              // constant buffer used for Mobius coefficients for CPU kernel
+  static __constant__ char mobius_d[size]; // buffer used for Mobius coefficients for GPU kernel
+
+  // helper trait for determining if we are using variable coefficients
+  template <Dslash5Type type> struct is_variable { static constexpr bool value = false; };
+  template <> struct is_variable<DSLASH5_MOBIUS_PRE> { static constexpr bool value = true; };
+  template <> struct is_variable<DSLASH5_MOBIUS> { static constexpr bool value = true; };
+  template <> struct is_variable<M5_INV_ZMOBIUS> { static constexpr bool value = true; };
 
   /**
-     @brief Helper function for grabbing the constant struct, whether
+     @brief Helper class for grabbing the constant struct, whether
      we are on the GPU or CPU.
   */
-  template <typename real> inline __device__ __host__ const coeff_5<real> *coeff()
+  template <typename real, bool is_variable, typename Arg> class coeff_type
   {
-#ifdef __CUDA_ARCH__
-    return reinterpret_cast<const coeff_5<real> *>(mobius_d);
-#else
-    return reinterpret_cast<const coeff_5<real> *>(mobius_h);
-#endif
-  }
-
-  template <typename real, Dslash5Type, typename Arg> struct coeff_type {
-    typedef real type;
+  public:
     const Arg &arg;
     __device__ __host__ coeff_type(const Arg &arg) : arg(arg) {}
     __device__ __host__ real a(int s) { return arg.a; }
@@ -41,12 +38,30 @@ namespace quda
     __device__ __host__ real c(int s) { return arg.c; }
   };
 
-  template <typename real, typename Arg> struct coeff_type<real, M5_INV_ZMOBIUS, Arg> {
-    typedef complex<real> type;
-    __device__ __host__ coeff_type(const Arg &arg) {}
-    __device__ __host__ complex<real> a(int s) { return coeff<real>()->a[s]; }
-    __device__ __host__ complex<real> b(int s) { return coeff<real>()->b[s]; }
-    __device__ __host__ complex<real> c(int s) { return coeff<real>()->c[s]; }
+  /**
+     @brief Specialization for variable complex coefficients
+  */
+  template <typename real, typename Arg> class coeff_type<real,true,Arg>
+  {
+    /**
+       @brief Helper function for grabbing the constant struct, whether
+       we are on the GPU or CPU.
+    */
+    inline __device__ __host__ const coeff_5<real> *coeff() const
+    {
+#ifdef __CUDA_ARCH__
+      return reinterpret_cast<const coeff_5<real> *>(mobius_d);
+#else
+      return arg.mobius_h;
+#endif
+    }
+
+  public:
+    const Arg &arg;
+    __device__ __host__ inline coeff_type(const Arg &arg) : arg(arg) {}
+    __device__ __host__ complex<real> a(int s) { return coeff()->a[s]; }
+    __device__ __host__ complex<real> b(int s) { return coeff()->b[s]; }
+    __device__ __host__ complex<real> c(int s) { return coeff()->c[s]; }
   };
 
   /**
@@ -76,6 +91,8 @@ namespace quda
 
     Dslash5Type type;
 
+    coeff_5<real> *mobius_h; // constant buffer used for Mobius coefficients for CPU kernel
+
     Dslash5Arg(ColorSpinorField &out, const ColorSpinorField &in, const ColorSpinorField &x, double m_f, double m_5,
         const Complex *b_5_, const Complex *c_5_, double a_, bool dagger, Dslash5Type type) :
         out(out),
@@ -97,10 +114,10 @@ namespace quda
         errorQuda("Unsupported field order out=%d in=%d\n", out.FieldOrder(), in.FieldOrder());
 
       if (sizeof(coeff_5<real>) > size) errorQuda("Coefficient buffer too large at %lu bytes\n", sizeof(coeff_5<real>));
-      coeff_5<real> *coeff = reinterpret_cast<coeff_5<real> *>(&mobius_h);
-      auto *a_5 = coeff->a;
-      auto *b_5 = coeff->b;
-      auto *c_5 = coeff->c;
+      mobius_h = new coeff_5<real>;
+      auto *a_5 = mobius_h->a;
+      auto *b_5 = mobius_h->b;
+      auto *c_5 = mobius_h->c;
 
       switch (type) {
       case DSLASH5_DWF: break;
@@ -151,6 +168,11 @@ namespace quda
 
       cudaMemcpyToSymbolAsync(mobius_d, mobius_h, sizeof(coeff_5<real>), 0, cudaMemcpyHostToDevice, streams[Nstream - 1]);
     }
+
+    virtual ~Dslash5Arg()
+    {
+      delete mobius_h;
+    }
   };
 
   /**
@@ -164,6 +186,7 @@ namespace quda
   __device__ __host__ inline void dslash5(Arg &arg, int parity, int x_cb, int s)
   {
     typedef typename mapper<Float>::type real;
+    coeff_type<real, is_variable<type>::value, Arg> coeff(arg);
     typedef ColorSpinor<real, nColor, 4> Vector;
 
     Vector out;
@@ -195,21 +218,19 @@ namespace quda
       out = x + arg.a * out;
     } else if (type == DSLASH5_MOBIUS_PRE) {
       Vector diagonal = arg.in(s * arg.volume_4d_cb + x_cb, parity);
-      auto *z = coeff<real>();
-      out = z->c[s] * out + z->b[s] * diagonal;
+      out = coeff.c(s) * out + coeff.b(s) * diagonal;
 
       if (xpay) {
         Vector x = arg.x(s * arg.volume_4d_cb + x_cb, parity);
-        out = x + z->a[s] * out;
+        out = x + coeff.a(s) * out;
       }
     } else if (type == DSLASH5_MOBIUS) {
       Vector diagonal = arg.in(s * arg.volume_4d_cb + x_cb, parity);
-      auto *z = coeff<real>();
-      out = z->c[s] * out + diagonal;
+      out = coeff.c(s) * out + diagonal;
 
       if (xpay) { // really axpy
         Vector x = arg.x(s * arg.volume_4d_cb + x_cb, parity);
-        out = z->a[s] * x + out;
+        out = coeff.a(s) * x + out;
       }
     }
 
@@ -268,7 +289,6 @@ namespace quda
   template <typename real, int nColor, bool dagger, Dslash5Type type, bool shared, typename Vector, typename Arg>
   __device__ __host__ inline Vector constantInv(Arg &arg, int parity, int x_cb, int s_)
   {
-
     const auto k = arg.b;
     const auto inv = arg.c;
 
@@ -326,10 +346,9 @@ namespace quda
   template <typename real, int nColor, bool dagger, Dslash5Type type, bool shared, typename Vector, typename Arg>
   __device__ __host__ inline Vector variableInv(Arg &arg, int parity, int x_cb, int s_)
   {
-
     constexpr int nSpin = 4;
     typedef ColorSpinor<real, nColor, nSpin / 2> HalfVector;
-    coeff_type<real, type, Arg> coeff(arg);
+    coeff_type<real, is_variable<type>::value, Arg> coeff(arg);
     Vector in = arg.in(s_ * arg.volume_4d_cb + x_cb, parity);
     Vector out;
 
@@ -411,7 +430,7 @@ namespace quda
     constexpr int nSpin = 4;
     typedef typename mapper<Float>::type real;
     typedef ColorSpinor<real, nColor, nSpin> Vector;
-    coeff_type<real, type, Arg> coeff(arg);
+    coeff_type<real, is_variable<type>::value, Arg> coeff(arg);
 
     Vector out;
     if (var_inverse) { // zMobius, must call variableInv

--- a/include/kernels/dslash_domain_wall_m5.cuh
+++ b/include/kernels/dslash_domain_wall_m5.cuh
@@ -19,10 +19,18 @@ namespace quda
   static __constant__ char mobius_d[size]; // buffer used for Mobius coefficients for GPU kernel
 
   // helper trait for determining if we are using variable coefficients
-  template <Dslash5Type type> struct is_variable { static constexpr bool value = false; };
-  template <> struct is_variable<DSLASH5_MOBIUS_PRE> { static constexpr bool value = true; };
-  template <> struct is_variable<DSLASH5_MOBIUS> { static constexpr bool value = true; };
-  template <> struct is_variable<M5_INV_ZMOBIUS> { static constexpr bool value = true; };
+  template <Dslash5Type type> struct is_variable {
+    static constexpr bool value = false;
+  };
+  template <> struct is_variable<DSLASH5_MOBIUS_PRE> {
+    static constexpr bool value = true;
+  };
+  template <> struct is_variable<DSLASH5_MOBIUS> {
+    static constexpr bool value = true;
+  };
+  template <> struct is_variable<M5_INV_ZMOBIUS> {
+    static constexpr bool value = true;
+  };
 
   /**
      @brief Helper class for grabbing the constant struct, whether
@@ -30,7 +38,7 @@ namespace quda
   */
   template <typename real, bool is_variable, typename Arg> class coeff_type
   {
-  public:
+public:
     const Arg &arg;
     __device__ __host__ coeff_type(const Arg &arg) : arg(arg) {}
     __device__ __host__ real a(int s) { return arg.a; }
@@ -41,7 +49,7 @@ namespace quda
   /**
      @brief Specialization for variable complex coefficients
   */
-  template <typename real, typename Arg> class coeff_type<real,true,Arg>
+  template <typename real, typename Arg> class coeff_type<real, true, Arg>
   {
     /**
        @brief Helper function for grabbing the constant struct, whether
@@ -56,7 +64,7 @@ namespace quda
 #endif
     }
 
-  public:
+public:
     const Arg &arg;
     __device__ __host__ inline coeff_type(const Arg &arg) : arg(arg) {}
     __device__ __host__ complex<real> a(int s) { return coeff()->a[s]; }
@@ -169,10 +177,7 @@ namespace quda
       cudaMemcpyToSymbolAsync(mobius_d, mobius_h, sizeof(coeff_5<real>), 0, cudaMemcpyHostToDevice, streams[Nstream - 1]);
     }
 
-    virtual ~Dslash5Arg()
-    {
-      delete mobius_h;
-    }
+    virtual ~Dslash5Arg() { delete mobius_h; }
   };
 
   /**

--- a/include/mpi_comm_handle.h
+++ b/include/mpi_comm_handle.h
@@ -8,9 +8,17 @@ extern MPI_Comm MPI_COMM_HANDLE;
 
 #ifdef QMP_COMMS
 #include <qmp.h>
+
+#ifdef __cplusplus
 extern "C" {
+#endif
+
 QMP_status_t QMP_get_mpi_comm(QMP_comm_t comm, void **mpicomm);
+
+#ifdef __cplusplus
 }
+#endif
+
 #endif
 
 #endif /* _COMM_HANDLE_H */

--- a/include/texture.h
+++ b/include/texture.h
@@ -22,14 +22,14 @@ template <typename OutputType, typename InputType> class Texture
     return *this;
   }
 
-  __device__ inline OutputType fetch(unsigned int idx)
+  __device__ inline OutputType fetch(unsigned int idx) const
   {
     OutputType rtn;
     copyFloatN(rtn, tex1Dfetch<RegType>(spinor, idx));
     return rtn;
   }
 
-  __device__ inline OutputType operator[](unsigned int idx) { return fetch(idx); }
+  __device__ inline OutputType operator[](unsigned int idx) const { return fetch(idx); }
 };
 
 __device__ inline double fetch_double(int2 v)
@@ -38,10 +38,10 @@ __device__ inline double fetch_double(int2 v)
 __device__ inline double2 fetch_double2(int4 v)
 { return make_double2(__hiloint2double(v.y, v.x), __hiloint2double(v.w, v.z)); }
 
-template <> __device__ inline double2 Texture<double2, double2>::fetch(unsigned int idx)
+template <> __device__ inline double2 Texture<double2, double2>::fetch(unsigned int idx) const
 { double2 out; copyFloatN(out, fetch_double2(tex1Dfetch<int4>(spinor, idx))); return out; }
 
-template <> __device__ inline float2 Texture<float2, double2>::fetch(unsigned int idx)
+template <> __device__ inline float2 Texture<float2, double2>::fetch(unsigned int idx) const
 { float2 out; copyFloatN(out, fetch_double2(tex1Dfetch<int4>(spinor, idx))); return out; }
 
 #else // !USE_TEXTURE_OBJECTS - use direct reads
@@ -68,7 +68,7 @@ template <typename OutputType, typename InputType> class Texture
     return *this;
   }
 
-  __device__ inline OutputType operator[](unsigned int idx)
+  __device__ inline OutputType operator[](unsigned int idx) const
   {
     OutputType out;
     copyFloatN(out, spinor[idx]);
@@ -207,7 +207,7 @@ template <typename RegType, typename StoreType, int N> class SpinorTexture
 
   virtual ~SpinorTexture() {}
 
-  __device__ inline void load(RegType x[], const int i, const int parity=0) {
+  __device__ inline void load(RegType x[], const int i, const int parity=0) const {
     // load data into registers first using the storage order
     constexpr int M = (N * vec_length<RegType>::value) / vec_length<InterType>::value;
     InterType y[M];
@@ -231,7 +231,7 @@ template <typename RegType, typename StoreType, int N> class SpinorTexture
      Load the ghost spinor.  For Wilson fermions, we assume that the
      ghost is spin projected
   */
-  __device__ inline void loadGhost(RegType x[], const int i, const int dim) {
+  __device__ inline void loadGhost(RegType x[], const int i, const int dim) const {
     // load data into registers first using the storage order
     const int Nspin = (N * vec_length<RegType>::value) / (3 * 2);
     // if Wilson, then load only half the number of components

--- a/include/texture.h
+++ b/include/texture.h
@@ -207,7 +207,8 @@ template <typename RegType, typename StoreType, int N> class SpinorTexture
 
   virtual ~SpinorTexture() {}
 
-  __device__ inline void load(RegType x[], const int i, const int parity=0) const {
+  __device__ inline void load(RegType x[], const int i, const int parity = 0) const
+  {
     // load data into registers first using the storage order
     constexpr int M = (N * vec_length<RegType>::value) / vec_length<InterType>::value;
     InterType y[M];
@@ -231,7 +232,8 @@ template <typename RegType, typename StoreType, int N> class SpinorTexture
      Load the ghost spinor.  For Wilson fermions, we assume that the
      ghost is spin projected
   */
-  __device__ inline void loadGhost(RegType x[], const int i, const int dim) const {
+  __device__ inline void loadGhost(RegType x[], const int i, const int dim) const
+  {
     // load data into registers first using the storage order
     const int Nspin = (N * vec_length<RegType>::value) / (3 * 2);
     // if Wilson, then load only half the number of components

--- a/lib/dslash_domain_wall_4d.cu
+++ b/lib/dslash_domain_wall_4d.cu
@@ -59,11 +59,11 @@ public:
       const auto kernel = DomainWall4DLaunch<void, 0, 0, 0, false, false, INTERIOR_KERNEL, Arg>::kernel;
       auto instance = Dslash<Float>::program_->kernel(kernel).instantiate(
         Type<Float>(), nDim, nColor, arg.nParity, arg.dagger, arg.xpay, arg.kernel_type, Type<Arg>());
-      cuMemcpyHtoDAsync(instance.get_constant_ptr("quda::mobius_d"), mobius_h, QUDA_MAX_DWF_LS * sizeof(complex<real>),
+      cuMemcpyHtoDAsync(instance.get_constant_ptr("quda::mobius_d"), arg.a_5, QUDA_MAX_DWF_LS * sizeof(complex<real>),
                         stream);
       Tunable::jitify_error = instance.configure(tp.grid, tp.block, tp.shared_bytes, stream).launch(arg);
 #else
-      cudaMemcpyToSymbolAsync(mobius_d, mobius_h, QUDA_MAX_DWF_LS * sizeof(complex<real>), 0, cudaMemcpyHostToDevice,
+      cudaMemcpyToSymbolAsync(mobius_d, arg.a_5, QUDA_MAX_DWF_LS * sizeof(complex<real>), 0, cudaMemcpyHostToDevice,
                               streams[Nstream - 1]);
       Dslash<Float>::template instantiate<DomainWall4DLaunch, nDim, nColor>(tp, arg, stream);
 #endif

--- a/lib/staggered_oprod.cu
+++ b/lib/staggered_oprod.cu
@@ -18,8 +18,7 @@ namespace quda {
 
   enum OprodKernelType { OPROD_INTERIOR_KERNEL, OPROD_EXTERIOR_KERNEL };
 
-  template<typename Float, typename Output, typename InputA, typename InputB>
-  struct StaggeredOprodArg {
+  template <typename Float, typename Output, typename InputA, typename InputB> struct StaggeredOprodArg {
     unsigned int length;
     const int parity;
     int dir;
@@ -35,9 +34,9 @@ namespace quda {
     unsigned int ghostOffset[4];
     bool partitioned[4];
 
-    StaggeredOprodArg(int parity, int dir, const unsigned int *ghostOffset,
-                      int displacement, const OprodKernelType &kernelType, int nFace, const double coeff[2],
-                      InputA &inA, InputB &inB, Output &outA, Output &outB, GaugeField &meta) :
+    StaggeredOprodArg(int parity, int dir, const unsigned int *ghostOffset, int displacement,
+                      const OprodKernelType &kernelType, int nFace, const double coeff[2], InputA &inA, InputB &inB,
+                      Output &outA, Output &outB, GaugeField &meta) :
       length(meta.VolumeCB()),
       parity(parity),
       dir(dir),
@@ -51,9 +50,9 @@ namespace quda {
     {
       this->coeff[0] = coeff[0];
       this->coeff[1] = coeff[1];
-      for(int i=0; i<4; ++i) this->X[i] = meta.X()[i];
-      for(int i=0; i<4; ++i) this->ghostOffset[i] = ghostOffset[i];
-      for(int i=0; i<4; ++i) this->partitioned[i] = commDimPartitioned(i) ? true : false;
+      for (int i = 0; i < 4; ++i) this->X[i] = meta.X()[i];
+      for (int i = 0; i < 4; ++i) this->ghostOffset[i] = ghostOffset[i];
+      for (int i = 0; i < 4; ++i) this->partitioned[i] = commDimPartitioned(i) ? true : false;
     }
   };
 
@@ -65,7 +64,7 @@ namespace quda {
   };
 
   template <IndexType idxType>
-  __device__ inline void coordsFromIndex(int& idx, int c[4], unsigned int cb_idx, int parity, const int X[4])
+  __device__ inline void coordsFromIndex(int &idx, int c[4], unsigned int cb_idx, int parity, const int X[4])
   {
       const int &LX = X[0];
       const int &LY = X[1];
@@ -121,48 +120,48 @@ namespace quda {
   
 
   // Get the  coordinates for the exterior kernels
-  __device__ inline void coordsFromIndex(int x[4], unsigned int cb_idx, const int X[4], int dir, int displacement, int parity)
-  {
-    int Xh[2] = {X[0]/2, X[1]/2};
-    switch(dir){
-    case 0:
-      x[2] = cb_idx/Xh[1] % X[2];
-      x[3] = cb_idx/(Xh[1]*X[2]) % X[3];
-      x[0] = cb_idx/(Xh[1]*X[2]*X[3]);
-      x[0] += (X[0] - displacement);
-      x[1] = 2*(cb_idx % Xh[1]) + ((x[0]+x[2]+x[3]+parity)&1);
-      break;
+    __device__ inline void coordsFromIndex(int x[4], unsigned int cb_idx, const int X[4], int dir, int displacement,
+                                           int parity)
+    {
+      int Xh[2] = {X[0] / 2, X[1] / 2};
+      switch (dir) {
+      case 0:
+        x[2] = cb_idx / Xh[1] % X[2];
+        x[3] = cb_idx / (Xh[1] * X[2]) % X[3];
+        x[0] = cb_idx / (Xh[1] * X[2] * X[3]);
+        x[0] += (X[0] - displacement);
+        x[1] = 2 * (cb_idx % Xh[1]) + ((x[0] + x[2] + x[3] + parity) & 1);
+        break;
 
-    case 1:
-      x[2] = cb_idx/Xh[0] % X[2];
-      x[3] = cb_idx/(Xh[0]*X[2]) % X[3];
-      x[1] = cb_idx/(Xh[0]*X[2]*X[3]);
-      x[1] += (X[1] - displacement);
-      x[0] = 2*(cb_idx % Xh[0]) + ((x[1]+x[2]+x[3]+parity)&1);
-      break;
+      case 1:
+        x[2] = cb_idx / Xh[0] % X[2];
+        x[3] = cb_idx / (Xh[0] * X[2]) % X[3];
+        x[1] = cb_idx / (Xh[0] * X[2] * X[3]);
+        x[1] += (X[1] - displacement);
+        x[0] = 2 * (cb_idx % Xh[0]) + ((x[1] + x[2] + x[3] + parity) & 1);
+        break;
 
-    case 2:
-      x[1] = cb_idx/Xh[0] % X[1];
-      x[3] = cb_idx/(Xh[0]*X[1]) % X[3];
-      x[2] = cb_idx/(Xh[0]*X[1]*X[3]);
-      x[2] += (X[2] - displacement);
-      x[0] = 2*(cb_idx % Xh[0]) + ((x[1]+x[2]+x[3]+parity)&1);
-      break;
+      case 2:
+        x[1] = cb_idx / Xh[0] % X[1];
+        x[3] = cb_idx / (Xh[0] * X[1]) % X[3];
+        x[2] = cb_idx / (Xh[0] * X[1] * X[3]);
+        x[2] += (X[2] - displacement);
+        x[0] = 2 * (cb_idx % Xh[0]) + ((x[1] + x[2] + x[3] + parity) & 1);
+        break;
 
-    case 3:
-      x[1] = cb_idx/Xh[0] % X[1];
-      x[2] = cb_idx/(Xh[0]*X[1]) % X[2];
-      x[3] = cb_idx/(Xh[0]*X[1]*X[2]);
-      x[3] += (X[3] - displacement);
-      x[0] = 2*(cb_idx % Xh[0]) + ((x[1]+x[2]+x[3]+parity)&1);
-      break;
-    }
-    return;
+      case 3:
+        x[1] = cb_idx / Xh[0] % X[1];
+        x[2] = cb_idx / (Xh[0] * X[1]) % X[2];
+        x[3] = cb_idx / (Xh[0] * X[1] * X[2]);
+        x[3] += (X[3] - displacement);
+        x[0] = 2 * (cb_idx % Xh[0]) + ((x[1] + x[2] + x[3] + parity) & 1);
+        break;
+      }
+      return;
   }
 
-
-  __device__ inline int neighborIndex(unsigned int cb_idx, const int shift[4], const bool partitioned[4],
-                                      int parity, const int X[4])
+  __device__ inline int neighborIndex(unsigned int cb_idx, const int shift[4], const bool partitioned[4], int parity,
+                                      const int X[4])
   {
     int full_idx;
     int x[4]; 
@@ -179,7 +178,6 @@ namespace quda {
     }
     return (((x[3]*X[2] + x[2])*X[1] + x[1])*X[0] + x[0]) >> 1;
   }
-
 
   template<typename real, typename Output, typename InputA, typename InputB>
   __global__ void interiorOprodKernel(StaggeredOprodArg<real, Output, InputA, InputB> arg)
@@ -360,51 +358,54 @@ namespace quda {
     comm_barrier();
   }
 
-  template<typename Float, typename Output, typename InputA, typename InputB>
-    void computeStaggeredOprodCuda(Output outA, Output outB, GaugeField& outFieldA, GaugeField& outFieldB, InputA& inA, InputB& inB, cudaColorSpinorField& src, 
-				   int parity, const int faceVolumeCB[4], const double coeff[2], int nFace)
-    {
-      unsigned int ghostOffset[4] = {0,0,0,0};
-      for(int dir=0; dir<4; ++dir) ghostOffset[dir] = src.GhostOffset(dir,1)/src.FieldOrder(); // offset we want is the forwards one
+  template <typename Float, typename Output, typename InputA, typename InputB>
+  void computeStaggeredOprodCuda(Output outA, Output outB, GaugeField &outFieldA, GaugeField &outFieldB, InputA &inA,
+                                 InputB &inB, cudaColorSpinorField &src, int parity, const int faceVolumeCB[4],
+                                 const double coeff[2], int nFace)
+  {
+    unsigned int ghostOffset[4] = {0, 0, 0, 0};
+    for (int dir = 0; dir < 4; ++dir)
+      ghostOffset[dir] = src.GhostOffset(dir, 1) / src.FieldOrder(); // offset we want is the forwards one
 
-      // Create the arguments for the interior kernel
-      StaggeredOprodArg<Float,Output,InputA,InputB> arg(parity, 0, ghostOffset, 1, OPROD_INTERIOR_KERNEL, nFace, coeff, inA, inB, outA, outB, outFieldA);
-      StaggeredOprodField<Float,Output,InputA,InputB> oprod(arg, outFieldA);
+    // Create the arguments for the interior kernel
+    StaggeredOprodArg<Float, Output, InputA, InputB> arg(parity, 0, ghostOffset, 1, OPROD_INTERIOR_KERNEL, nFace, coeff,
+                                                         inA, inB, outA, outB, outFieldA);
+    StaggeredOprodField<Float, Output, InputA, InputB> oprod(arg, outFieldA);
 
-      arg.kernelType = OPROD_INTERIOR_KERNEL;
-      arg.length = src.VolumeCB();
-      oprod.apply(streams[Nstream-1]); 
+    arg.kernelType = OPROD_INTERIOR_KERNEL;
+    arg.length = src.VolumeCB();
+    oprod.apply(streams[Nstream - 1]);
 
-      for(int i=3; i>=0; i--){
-	if (commDimPartitioned(i)) {
-	  // update parameters for this exterior kernel
-	  arg.kernelType = OPROD_EXTERIOR_KERNEL;
-	  arg.dir = i;
+    for (int i = 3; i >= 0; i--) {
+      if (commDimPartitioned(i)) {
+        // update parameters for this exterior kernel
+        arg.kernelType = OPROD_EXTERIOR_KERNEL;
+        arg.dir = i;
 
-	  // First, do the one hop term
-	  {
-	    arg.displacement = 1;
-	    arg.length = faceVolumeCB[i];
-	    oprod.apply(streams[Nstream-1]);
-	  }
+        // First, do the one hop term
+        {
+          arg.displacement = 1;
+          arg.length = faceVolumeCB[i];
+          oprod.apply(streams[Nstream - 1]);
+        }
 
-	  // Now do the 3 hop term
-	  if (nFace == 3) {
-	    arg.displacement = 3;                      
-	    arg.length = arg.displacement*faceVolumeCB[i];
-	    oprod.apply(streams[Nstream-1]);
-	  } 
-	}
-      } // i=3,..,0
+        // Now do the 3 hop term
+        if (nFace == 3) {
+          arg.displacement = 3;
+          arg.length = arg.displacement * faceVolumeCB[i];
+          oprod.apply(streams[Nstream - 1]);
+        }
+      }
+    } // i=3,..,0
 
-      checkCudaError();
+    checkCudaError();
     } // computeStaggeredOprodCuda
 
 #endif // GPU_STAGGERED_DIRAC
 
-  void computeStaggeredOprod(GaugeField& outA, GaugeField& outB, ColorSpinorField& inEven, ColorSpinorField& inOdd,
-			     int parity, const double coeff[2], int nFace)
-  {
+    void computeStaggeredOprod(GaugeField &outA, GaugeField &outB, ColorSpinorField &inEven, ColorSpinorField &inOdd,
+                               int parity, const double coeff[2], int nFace)
+    {
 #ifdef GPU_STAGGERED_DIRAC
     if(outA.Order() != QUDA_FLOAT2_GAUGE_ORDER)
       errorQuda("Unsupported output ordering: %d\n", outA.Order());    

--- a/lib/staggered_oprod.cu
+++ b/lib/staggered_oprod.cu
@@ -19,42 +19,42 @@ namespace quda {
   enum OprodKernelType { OPROD_INTERIOR_KERNEL, OPROD_EXTERIOR_KERNEL };
 
   template<typename Float, typename Output, typename InputA, typename InputB>
-    struct StaggeredOprodArg {
-      unsigned int length;
-      int X[4];
-      unsigned int parity;
-      unsigned int dir;
-      unsigned int ghostOffset[4];
-      unsigned int displacement;
-      OprodKernelType kernelType;
-      int nFace;
-      bool partitioned[4];
-      InputA inA;
-      InputB inB;
-      Output outA;
-      Output outB;
-      Float coeff[2];
+  struct StaggeredOprodArg {
+    unsigned int length;
+    const int parity;
+    int dir;
+    int displacement;
+    OprodKernelType kernelType;
+    const int nFace;
+    const InputA inA;
+    const InputB inB;
+    Output outA;
+    Output outB;
+    Float coeff[2];
+    int X[4];
+    unsigned int ghostOffset[4];
+    bool partitioned[4];
 
-      StaggeredOprodArg(const unsigned int parity, const unsigned int dir, const unsigned int *ghostOffset,
-          const unsigned int displacement, const OprodKernelType &kernelType, const int nFace, const double coeff[2],
-          InputA &inA, InputB &inB, Output &outA, Output &outB, GaugeField &meta) :
-          length(meta.VolumeCB()),
-          parity(parity),
-          dir(dir),
-          displacement(displacement),
-          kernelType(kernelType),
-          nFace(nFace),
-          inA(inA),
-          inB(inB),
-          outA(outA),
-          outB(outB)
-      {
-        this->coeff[0] = coeff[0];
-        this->coeff[1] = coeff[1];
-        for(int i=0; i<4; ++i) this->X[i] = meta.X()[i];
-        for(int i=0; i<4; ++i) this->ghostOffset[i] = ghostOffset[i];
-        for(int i=0; i<4; ++i) this->partitioned[i] = commDimPartitioned(i) ? true : false;
-      }
+    StaggeredOprodArg(int parity, int dir, const unsigned int *ghostOffset,
+                      int displacement, const OprodKernelType &kernelType, int nFace, const double coeff[2],
+                      InputA &inA, InputB &inB, Output &outA, Output &outB, GaugeField &meta) :
+      length(meta.VolumeCB()),
+      parity(parity),
+      dir(dir),
+      displacement(displacement),
+      kernelType(kernelType),
+      nFace(nFace),
+      inA(inA),
+      inB(inB),
+      outA(outA),
+      outB(outB)
+    {
+      this->coeff[0] = coeff[0];
+      this->coeff[1] = coeff[1];
+      for(int i=0; i<4; ++i) this->X[i] = meta.X()[i];
+      for(int i=0; i<4; ++i) this->ghostOffset[i] = ghostOffset[i];
+      for(int i=0; i<4; ++i) this->partitioned[i] = commDimPartitioned(i) ? true : false;
+    }
   };
 
   enum IndexType {
@@ -65,8 +65,7 @@ namespace quda {
   };
 
   template <IndexType idxType>
-    static __device__ __forceinline__ void coordsFromIndex(int& idx, int c[4],  
-        const unsigned int cb_idx, const unsigned int parity, const int X[4])
+  __device__ inline void coordsFromIndex(int& idx, int c[4], unsigned int cb_idx, int parity, const int X[4])
   {
       const int &LX = X[0];
       const int &LY = X[1];
@@ -122,7 +121,7 @@ namespace quda {
   
 
   // Get the  coordinates for the exterior kernels
-  __device__ static void coordsFromIndex(int x[4], const unsigned int cb_idx, const int X[4], const unsigned int dir, const int displacement, const unsigned int parity)
+  __device__ inline void coordsFromIndex(int x[4], unsigned int cb_idx, const int X[4], int dir, int displacement, int parity)
   {
     int Xh[2] = {X[0]/2, X[1]/2};
     switch(dir){
@@ -162,13 +161,13 @@ namespace quda {
   }
 
 
-  __device__ __forceinline__
-  int neighborIndex(const unsigned int cb_idx, const int shift[4],  const bool partitioned[4], const unsigned int parity, const int X[4]){
+  __device__ inline int neighborIndex(unsigned int cb_idx, const int shift[4], const bool partitioned[4],
+                                      int parity, const int X[4])
+  {
     int full_idx;
     int x[4]; 
 
     coordsFromIndex<EVEN_X>(full_idx, x, cb_idx, parity, X);
-
     
     for(int dim = 0; dim<4; ++dim){
       if( partitioned[dim] )
@@ -363,7 +362,7 @@ namespace quda {
 
   template<typename Float, typename Output, typename InputA, typename InputB>
     void computeStaggeredOprodCuda(Output outA, Output outB, GaugeField& outFieldA, GaugeField& outFieldB, InputA& inA, InputB& inB, cudaColorSpinorField& src, 
-				   const unsigned int parity, const int faceVolumeCB[4], const double coeff[2], int nFace)
+				   int parity, const int faceVolumeCB[4], const double coeff[2], int nFace)
     {
       unsigned int ghostOffset[4] = {0,0,0,0};
       for(int dir=0; dir<4; ++dir) ghostOffset[dir] = src.GhostOffset(dir,1)/src.FieldOrder(); // offset we want is the forwards one
@@ -404,7 +403,7 @@ namespace quda {
 #endif // GPU_STAGGERED_DIRAC
 
   void computeStaggeredOprod(GaugeField& outA, GaugeField& outB, ColorSpinorField& inEven, ColorSpinorField& inOdd,
-			     const unsigned int parity, const double coeff[2], int nFace)
+			     int parity, const double coeff[2], int nFace)
   {
 #ifdef GPU_STAGGERED_DIRAC
     if(outA.Order() != QUDA_FLOAT2_GAUGE_ORDER)

--- a/tests/invertmsrc_test.cpp
+++ b/tests/invertmsrc_test.cpp
@@ -545,9 +545,8 @@ int main(int argc, char **argv)
       } else if (dslash_type == QUDA_DOMAIN_WALL_4D_DSLASH) {
         dw_4d_matpc(spinorCheck[i], gauge, spinorOutMulti[i], kappa5, inv_param.matpc_type, 0, inv_param.cpu_prec, gauge_param, inv_param.mass);
       } else if (dslash_type == QUDA_MOBIUS_DWF_DSLASH) {
-        double *kappa_b, *kappa_c;
-        kappa_b = (double*)malloc(Lsdim*sizeof(double));
-        kappa_c = (double*)malloc(Lsdim*sizeof(double));
+        double _Complex *kappa_b = (double _Complex *)malloc(Lsdim * sizeof(double _Complex));
+        double _Complex *kappa_c = (double _Complex *)malloc(Lsdim * sizeof(double _Complex));
         for(int xs = 0 ; xs < Lsdim ; xs++)
         {
           kappa_b[xs] = 1.0/(2*(inv_param.b_5[xs]*(4.0 + inv_param.m5) + 1.0));

